server/src/main/java/org/opensearch/identity/Subject.java

```diff
@@ -29,4 +30,11 @@ public interface Subject {
      * throws SubjectDisabled
      */
     void authenticate(final AuthToken token);
+
+    /**
+     * Checks scopes of the current subject if they are allowed for any of the listed scopes
+     * @param scope The scopes to check against the subject
+     * @return true if allowed, false if none of the scopes are allowed.
+     */
+    boolean isAllowed(final List<Scope> scope);
```

Comment:
I don't think we should accept alternative implementations for this method, lets change around this update to support application scopes.  While there are some parts of OpenSearch that should be extensible/configurable how applications are allowed to access high-grain functionality seems better served consistantly.

I recommend is instead create a class in OpenSearch, ApplicationAwareSubject, this class would implement Subject via wrapping an existing subject from an Identity Plugin.  This new class would be final, and it would wrap the subject in the IdentityService class.

Then have Subject include a method, `Principal getApplication()` that would be all of the application associated with the account.  This would allow onBehalfOf, app tokens, services accounts, and other features specific to IdentityPlugins to pass application reference if it has one.

Then the isAllowed(...) check is only implemented in ApplicationAwareSubject, even if the NoopIdentityPlugin is used application scopes are still required

Reply:
Hi @peternied, thank you for the suggestion. That sounds like a good idea. I am not sure I follow the third paragraph however. 

I am not sure what you mean by "applications associated with the account." If we have ApplicationAwareSubject class, we can then add that method you suggested but we have not been associating Extensions with a single subject. Perhaps you are suggesting that we make each extension create its own ApplicationAwareSubject which can track the extension it is associated with? I am also not sure what you had in mind for the application reference. 

I also don't totally agree that we want to have scopes active even if the NoopIdentityPlugin is used. It seems like this would be counterintuitive to the majority of the Noop implementations which all accept everything. In this case, I would expect the scopes to be disregarded. Maybe you could explain why you feel otherwise?

Reply:
Checkout this commit https://github.com/peternied/OpenSearch-1/commit/128e531a4020b3e29b78b735079c146180eaadf0 that shows how this could be done, it might be easier to see the implementation, feel free to merge that commit into this PR or riff on it as you'd like